### DISCUSSION TITLE
Mock API call during testing

### DISCUSSION THREAD

**Discussion by notAro14** [trpc/trpc#1879] [Q&A] [answered]
Hello, is it possible to mock API call with [Mock Service Worker](https://mswjs.io/) during integration testing in `jest` and `testing-library` ?

**Comment by KATT** [comment, marked as answer]
Don't mock api calls that you have full control of, control what's in the db instead.


See this for inspo https://github.com/trpc/examples-next-prisma-starter/blob/main/src/server/routers/post.test.ts

**Reply by ChristianIvicevic** [reply]
I do not agree with this suggestion at all. This specifically requires to somehow boot up a database in the CI pipeline just for running tests making this process needlessly complex and heavy. There needs to be a more convenient way to test components that are using trpc.

**Reply by KATT** [reply]
You are free to do what you want. This is my advice.

Spinning up a db is neither hard nor adds significant overhead to the test duration whilst making your tests be a lot closer to reality

**Reply by KATT** [reply]
Fwiw, the official starter projects from trpc already have it set up for you

**Reply by ChristianIvicevic** [reply]
The sheer setup is still convoluted. Suppose I am using a production database on PlanetScale and developing a local feature branch of my application against a branched version of the production database. Once I am finished with the raw implementation of the feature I move on to writing tests. Instead of using something battle-tested such as msw, which is highly popular and very common in conventional cases of working with REST API, you now propose to run the dev server of Next.js against an actual local database.

First, this requires two separate settings or variables to instruct Next.js to connect to the branched database on PlanetScale or, say, a local database that is solely used for Jest. By doing the latter it is now impossible to have a workflow of having the local branch running against the branched database while also running tests in parallel. Something many do in order to instantly see their changes in the browser and have the confidence the tests are still green.

Second, tests cannot be run in parallel anymore as you want to ensure they each run in isolation and thus need a strictly isolated database. As such your claim that working on a real database doesn't add to the test duration is factually incorrect when you have the potential to run tests in parallel, if they are isolated accordingly.

Third, business rules often have complex requirements. As such preparing a local database used during testing with the correct data is cumbersome. Think of multiple entities required in various tables with correct relations. For what? Maybe you just want to test whether a component correctly displays fetched data. A unit test should be lightweight and should not depend on the database setup or representation when all I am interested in is that I want to confirm the component behaves in a specific way when it receives a certain response from the API such as successful responses or erroneous ones.

Lastly, what you propose is almost a classic e2e setup in tandem with Cypress or Playwright, which I'd be using for smoke tests or bigger tests to ensure features still work. However for more common frontend tests which are component tests after all I highly disagree with your proposal.

With all of this being said, I do understand that we're at and impasse and won't come to an agreement regarding testing strategies. Be that as it may, at this point I want to you to know, as long as components using tRPC cannot be tested properly in isolation and my code ends up untested I will not be using tRPC and will gladly return back to classic REST API communication with `react-query` until tRPC has improved in that regard even if it means sacrificing a bit of DX.

In short: I want to test my UI code and should not require to boot up server and database for that.

**Reply by mmkal** [reply]
Why can't you just use msw with trpc as is? trpc/server is basically just an express middleware which handles GET and POST results. The query parameters and response shapes are a little different than you'd have in a normal server, but otherwise it's the same as a "classic" REST setup.

**Reply by ChristianIvicevic** [reply]
Unfortunately due to the protocol which it communicates with it requires some more complex handling. I was actually able to handle tRPC requests with msw successfully but those can be batched which requires splitting them up and responding in a similar fashion. It is definitely possible but requires some boilerplate which I don't want to spend too much time on writing.

**Reply by ChristianIvicevic** [reply]
Here is an initial concept of what is necessary to handle requests with msw:
```typescript
import { render, screen, waitFor } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import nodeFetch from 'node-fetch';
import queryString from 'query-string';
import type { ReactNode } from 'react';
import { useState } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { trpc } from '../utils/trpc';
import { Sample } from './sample';

// According to JSON-RPC 2.0 and tRPC documentation.
// https://trpc.io/docs/rpc
const jsonSucessResponse = (data: unknown) => ({
  id: null,
  result: { type: 'data', data },
});

// Single request example URL
// /api/trpc/hello?batch=1&input=%7B%220%22%3A%7B%22text%22%3A%22User%22%7D%7D
const sampleHandler = rest.get(
  'http://mock-api/api/trpc/*',
  (req, res, ctx) => {
    const operations = Object.values(req.params);
    const { input } = queryString.parse(req.url.search);

    // TODO: Handle each operation separately wherein the n-th operation has its
    //  input as the n-th element of the input array and return an array of
    //  responses.

    return res(
      ctx.status(200),
      ctx.json([jsonSucessResponse({ greeting: 'Hello, MSW!' })]),
    );
  },
);

const mswServer = setupServer(...[sampleHandler]);

const Wrapper = ({ children }: { children: ReactNode }) => {
  const [queryClient] = useState(() => new QueryClient());
  const [trpcClient] = useState(() =>
    trpc.createClient({
      url: 'http://mock-api/api/trpc',
      fetch: nodeFetch as unknown as typeof fetch,
    }),
  );

  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </trpc.Provider>
  );
};

describe('Sample', () => {
  beforeAll(() => mswServer.listen());

  afterAll(() => mswServer.close());

  afterEach(() => mswServer.resetHandlers());

  it('should render successfully', async () => {
    const { baseElement } = render(<Sample />, { wrapper: Wrapper });
    await waitFor(() => screen.findByText(/MSW/));
    expect(baseElement).toBeTruthy();
  });
});
```
You might get the gist of it. It will surely be possible to introduce helper functions to reduce the amount of boilerplate. Another nice to have is to deduce all available endpoints and supply custom handlers via some form of `Record<Endpoints, Handler>` where `Endpoints` is something like
```typescript
type Endpoints = {
  [K in keyof AppRouter['_def']['queries']]: K;
}[keyof AppRouter['_def']['queries']];
```

**Reply by mmkal** [reply]
Honestly, that doesn't look too bad to me. The whole `<Wrapper>` component and msw boilerplate can be moved to a module, you only need to write it once and it can be reused for all your component tests. Or you could open source your own trpc-msw package - I happen to share @KATT's unenthusiasm for mocking, but you will surely find many others who find it useful, and who find it fits with their project's existing approach to testing. I doubt you'll persuade @KATT to make it part of trpc though!

A few things you might want to tweak from that example before it'd be ready for whatever you threw at it:

- disable batching to make request mocking easier
- handle mutations too (actually easier, they're just regular post requests when not batched)
- maybe export the beforeAll, afterAll and beforeEach helpers?
- if you're using zod and output schemas, you could even pipe them through zod-to-json-schema and json-schema-faker to autogenerate mocked responses. 

Other than that, it looks fine to me as is!

**Reply by ChristianIvicevic** [reply]
> disable batching to make request mocking easier

While we're at that topic, do you happen to know how to disable batching in that case? I honestly didn't find an option when creating the client like in my snippet unlike the Next.js integration where it's an actual configurable option.

**Reply by KATT** [reply]
I test extremely complex business logic, in parallel & with real databases, and guess what? It works **great** without mocking. I don't use PlanetScale, but my setup would look exactly the same if I did. The only thing I only ever mock is the third-party SDKs which I have no choice but to mock.

If we've managed to create a central piece of tech for your application to work, maybe, just maybe, we kinda know what we're doing and there could be something to learn here? As a reminder, this is free open-source software which you haven't contributed to in any shape or form and we don't owe you anything, try to be empathetic & respectful when we are helping you out for free.

If you want to skip batching you can set up your client with a `httpLink` instead of a `httpBatchLink` - it's in the docs: https://trpc.io/docs/links#disabling-request-batching.

**Reply by moltar** [reply]
There are many use cases for mocking beyond "writing to a database.

Two that immediately come to mind for our use case:

1. Using a mock API for stories (Storybook, Ladle). For this use case, even if we could use a real endpoint, it wouldn't be a good idea, as one generally wants to control the data and not have some gibberish mock data created by test users.
2. Endpoints that go beyond database calls. We have a use case that requires a tight integration with many AWS services, including ones that would bill you a lot, if you were just creating these resources for testing. In this case also running a live API isn't even feasible locally. And deploying an actual API can take many minutes. So you can't even tear it down for each test. 

I agree on integration > mocking premise in general. But there are _some_ cases that are just not feasible without having mocks.

**Reply by KATT** [reply]
True that it'd be nice for storybook & probably not that hard to make a helper where you could inject a mock router

**Reply by moltar** [reply]
@KATT If you could throw together a verbal sketch of your ideal solution, then I can see if we can allocate internal resources for implementing this. Thanks.

**Reply by uncvrd** [reply]
@ChristianIvicevic thanks for leading me in the right direction with typing an `Endpoints` object to get the keys, I found a solution on stackoverflow that will recursively join all keys together with periods but seems a bit excessive haha, did you end up finding a different approach for accessing all "endpoints" via string in your project?

Heres that SO solution: https://stackoverflow.com/a/58436959

I can then use that type as a parameter to intercept batch http requests that `include` the endpoint query

**Reply by KATT** [reply]
FWIW, @moltar & @notAro14, I may have gone a bit too hard against this one.

I won't build stuff around mocking as I don't need it myself, but if people want to contribute, I'd be open to people's contributions.

**Reply by airtonix** [reply]
For everyone looking for the real answer: https://github.com/maloguertin/msw-trpc

**Reply by swport** [reply]
Either msw-trpc has poor documentation or it isn't designed to test components which are using trpc api

**Comment by notAro14** [comment]
> Don't mock api calls that you have full control of, control what's in the db instead.
> 
> See this for inspo https://github.com/trpc/examples-next-prisma-starter/blob/main/src/server/routers/post.test.ts

I did not try yet but just to be sure, you mean that I have to write directly to the DB during testing ? It's a bit unusual to me...

**Reply by KATT** [reply]
You'll thank me later

**Reply by airtonix** [reply]
no we won't.

evidentally, many years later we didn't

**Comment by martin-fv** [comment]
Here's an example implementation for Storybook with a typesafe response.
https://gist.github.com/martin-fv/9bad75363245a16c918e112eb6a19f0d

**Reply by mwarger** [reply]
This is awesome.  I am trying this out and seeing an error, though.  

`Query data cannot be undefined. Please make sure to return a value other than undefined from your query function.`

I'm wondering if you have seen this?  While inspecting things, it seems like I have data for all my queries.  I'll try to reproduce this in a public repo...

**Reply by mwarger** [reply]
@martin-fv I actually figured this out - I had `superjson` enabled as a transform.  It seems like this changes the output in a way that breaks this approach.  Disabling superjson fixed this problem.

**Reply by martin-fv** [reply]
Good to know, yes we don't use superjson. Also there's been changes in the last rc, I will post an update soon.

**Reply by mwarger** [reply]
I managed to get this working as well with superjson by adding a call to serialize to the jsonRpcSuccessResponse.

```
export const jsonRpcSuccessResponse = (data: unknown) => {
  const response = superjson.serialize(data);

  return {
    id: null,
    result: { type: 'data', data: response },
  };
};
```

**Reply by mwarger** [reply]
> Good to know, yes we don't use superjson. Also there's been changes in the last rc, I will post an update soon.

Interesting - I thought I was on the latest, but perhaps not.  Does this approach break in the latest version?

**Reply by martin-fv** [reply]
Changed GetInferenceHelpers to RouterInput, RouterOutput due to the latest TRPC changes, see gist above. I also added an example of supertest.

**Reply by uncvrd** [reply]
@martin-fv thank you for this gist, this gets me on the right track. I have some queries that are three layers deep and I'm having some problems understanding the typescript that's used to infer the types and unfortunately I cannot just add another generic to go one level deeper like this:

```
K3 extends keyof RouterInput[K2]
```

Would you happen to know how to accomplish this? Thank you very much!

**Reply by martin-fv** [reply]
I've had a quick look @uncvrd but didn't get it working. I'm not that familiar with generics.
@KATT in v9 we were able to get all the routes in one type (ex. "user.post.create" | "user.post.delete") is there are a way to achieve in the new version? Since my current approach is limited to 2-level nested routes.

**Reply by uncvrd** [reply]
@martin-fv no problem - thanks for looking in to it :) I struggle with generics as well. I do have a version that uses some typescript magic that I linked above from a SO question, this is what I currently have:

```
// helper.ts

type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
    11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...0[]]

type Join<K, P> = K extends string | number ?
    P extends string | number ?
    `${K}${"" extends P ? "" : "."}${P}`
    : never : never;

type Paths<T, D extends number = 3> = [D] extends [never] ? never : T extends object ?
    { [K in keyof T]-?: K extends string | number ?
        `${K}` | Join<K, Paths<T[K], Prev[D]>>
        : never
    }[keyof T] : ""

    
export const endpointIncludes = (url: string, query: Paths<RouterInput, 3>) => {
    return url.includes(query);
};
```

Where I can then use `endpointIncludes` to intercept routes that include the specific "string joined" paths:

```
// auth.cy.ts

cy.intercept("POST", "/api/trpc/*", (req) => {
    if (endpointIncludes(req.url, "authed.organization.updateBillingContact")) {
        req.alias = "update";
    }
})
```

This just feels wrong, but it's the best I can do for now! I would love to be able to have a better way to strongly type my intercepts so that when I (inevitably) change my routes I can catch them easily in these files as well 🙏

**Reply by sneko** [reply]
@martin-fv you saved me some time for sure! Thank you :)

> with a typesafe response.

What did you mean? According to my test I can set anything inside the "response" property, it won't say it's not possible according the the trpc app router.

**Reply by martin-fv** [reply]
It means the response should be typed, so you don't autocomplete there?

Try to use Trpc version 10.0.0, I think there are some breaking changes in the last version for this script.

**Reply by airtonix** [reply]
https://gist.github.com/martin-fv/9bad75363245a16c918e112eb6a19f0d?permalink_comment_id=4466216#gistcomment-4466216

**Comment by testerez** [comment]
I'm trying a different approach, using `setData` to populate query cache before the story is rendered. In case anyone is interested, here is my code (to be refactored so that it can be easily used in any story):

```ts
import type { Meta } from '@storybook/react'
import { testFixtures } from 'src/utils/test/testUtils'
import { PropsWithChildren, useState } from 'react'
import { createTRPCReact, httpBatchLink } from '@trpc/react-query'
import { AppRouter } from 'src/server/trpc/router'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

// The component to be showcased
const Post = () => {
  const query = trpc.post.byId.useQuery({ id: `123` })
  return (
    <>
      <div>Status: {query.status}</div>
      <pre>data: {JSON.stringify(query.data, null, 2)}</pre>
    </>
  )
}

const trpc = createTRPCReact<AppRouter>()
const StorybookTrpcProvider = ({ children }: PropsWithChildren) => {
  const [queryClient] = useState(
    new QueryClient({ defaultOptions: { queries: { staleTime: Infinity } } }),
  )
  const [trpcClient] = useState(() =>
    trpc.createClient({
      links: [httpBatchLink({ url: '' })],
    }),
  )
  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>
        {children}
      </QueryClientProvider>
    </trpc.Provider>
  )
}

type TrpcContext = ReturnType<typeof trpc['useContext']>

// Hack to be able to access trpcContext
const ActOnTrpcContext = ({
  callback,
  children,
}: PropsWithChildren<{
  callback: (trpcContext: TrpcContext) => void
}>) => {
  const trpcContext = trpc.useContext()
  callback(trpcContext)
  return <>{children}</>
}

const withTrpcContext =
  (callback: (context: TrpcContext) => void) => (Story: any) =>
    (
      <ActOnTrpcContext callback={callback}>
        <Story />
      </ActOnTrpcContext>
    )

const meta: Meta<typeof Post> = {
  title: 'Post',
  component: Post,
  decorators: [
    (Story) => (
      <StorybookTrpcProvider>
        <Story />
      </StorybookTrpcProvider>
    ),
  ],
}
export default meta

export const UsdWithNoPlugins = {
  render: () => <Post />,
  decorators: [
    withTrpcContext((c) => {
      c.post.byId.setData({ id: '123' }, () => testFixtures.post)
    }),
  ],
}

```

**Reply by sneko** [reply]
Thought about this today but didn't find the right way to set the data (was fighting  with other methods :/ ). In the meantime I adopted the @martin-fv way, but yours is quite cool because it avoids using a third-party tool to mock network calls. Congrats ;)

**Reply by sneko** [reply]
@testerez , just thinking out loud, does your solution can be adapted to mutations? From what I remember they don't expose the same `defaultOptions` properties.

**Reply by testerez** [reply]
If you need to mock mutations data, I guess `msw` is a better avenue.

**Comment by lvpaspaltzis** [comment]
I created a lib to intercept tRPC calls using **Cypress**: https://github.com/vargasmesh/cy-trpc. 

I came to this discussion looking for a way to mock tRPC, but in my case, I had to do it in Cypress component tests. Although the `msw` solution works great, I couldn't afford to add a new testing library only for tRPC. So to promote tRPC adoption for anyone using Cypress, I created a dynamic stub for it (highly inspired by [msw-trpc](https://github.com/maloguertin/msw-trpc))

**Comment by swport** [comment]
good job at making me hate testing even more..
